### DISCUSSION TITLE
Path to TypeScript strict mode

### DISCUSSION THREAD

**Discussion by Manc** [LegendApp/legend-state#53] [General]
Hi @jmeistrich,

it's probably best to continue #42 here in a discussion as I may have more questions which are no longer really related to that ticket.

So my first questions are about `NodeValue`:

- What is the `proxy` property (currently an optional unspecified `object`)? Is there already a type anywhere or should a new one be created? Is there a particular reason why it's `object`?
- What does `getNodeValue` do or return? The argument of `getNodeValue` is a `NodeValue` and returns `any`. Maybe this can be made a bit clearer.

**Comment by Manc** [comment]
One more question… We already briefly [talked](https://github.com/LegendApp/legend-state/issues/42#issuecomment-1283878889) about class vs prototype object (I think this could also be named ES6 class vs ES5 class).

The only reason why I'm bringing this up again is because I didn't know how to make proper types for it. Do you remember what the performance difference was? Did you read about it or did you even have a benchmark test or something for this library?

I genuinely don't know about the performance difference, but searching online I stumbed across a StackOverflow question about this and the theory that a class was slower was kindof dismissed – at least it depends on the specifics maybe.

So I tried to do a [comparison on jsben.ch](https://jsben.ch/PjiFU) and the result is that both are pretty much identical (I ran this with Firefox and Safari). I don't know how realistic my benchmark is as it is obviously a simplified version of the code and the test is basically, creating a lot of new instances and then calling a function on it.

Maybe there have been advancements in browsers how well ES6 classes are supported. So, by no means do I want to say that the `prototype` way is wrong, but maybe you can shed some light on it.

**Reply by jmeistrich** [reply]
Oh that benchmark is fascinating! I think the main reason to keep it as a prototype is that it results in a smaller file size. I've been trying to keep the core library around 3kb and that brought it down a couple hundred bytes when compiling it with esbuild (tested with `npm run checksize`). But of course I could be wrong about that - it's possible that different compiler settings would have different results?

I tracked down the place I saw typing a function with prototype as a class: https://github.com/preactjs/signals/blob/main/packages/core/src/index.ts#L160. So I think if we keep it as a prototype but export types as if it's a class, we get the best of both worlds?

**Comment by jmeistrich** [comment]
Good idea. I'll close #42 and leave a comment there to see this discussion.

- The `proxy` property contains the `Proxy` for that node. It's created in `getProxy`. I think I made that type way in the early days and didn't know what to do with it, but it seems like the type of a Proxy is the same as the wrapped value, so it can actually be `NodeValue`.
- `getNodeValue` is the meat of the whole library 😅. It gets the raw value at a given node. I think to type it correctly we'd have to make NodeValue generic with the type of the value at the node, but it's also possible that can change... For the purposes of the internals it doesn't really matter what that type is, as long as the public-facing APIs return the right type? Would typing it as `unknown` be better? Or would it be better to try to pipe the actual types through?

**Comment by jmeistrich** [comment]
Hey @Manc, I've been working on the types in the past couple weeks, following your lead. How is it looking now from your perspective?
